fml/platform/fuchsia/message_loop_fuchsia.cc

```diff
@@ -11,28 +11,52 @@
 namespace fml {
 
 MessageLoopFuchsia::MessageLoopFuchsia()
-    : loop_(&kAsyncLoopConfigAttachToCurrentThread) {}
+  : loop_(&kAsyncLoopConfigAttachToCurrentThread),
+    running_(false) {
+  auto status = zx_timer_create(ZX_TIMER_SLACK_LATE, ZX_CLOCK_MONOTONIC, &timer_);
+  FML_CHECK(status == ZX_OK);
+  timer_wait_ = new async::Wait(timer_, ZX_TIMER_SIGNALED);
+
+  auto wait_handler = [&](async_dispatcher_t* dispatcher,
+                        async::Wait* wait,
+                        zx_status_t status,
+                        const zx_packet_signal_t* signal
+                    ) {
+    RunExpiredTasksNow();
+    wait->Begin(loop_.dispatcher());
+  };
+
+  timer_wait_->set_handler(wait_handler);
+  status = timer_wait_->Begin(loop_.dispatcher());
+  FML_CHECK(status == ZX_OK);
+}
 
-MessageLoopFuchsia::~MessageLoopFuchsia() = default;
+MessageLoopFuchsia::~MessageLoopFuchsia() {
+  delete timer_wait_;
+  zx_handle_close(timer_);
+}
 
 void MessageLoopFuchsia::Run() {
-  loop_.Run();
+  running_ = true;
+
+  while (running_) {
+    auto status = loop_.Run(zx::time::infinite(), true);
+
+    // This handles the case where the loop is terminated using zx APIs.
+    if (status == ZX_ERR_CANCELED) {
+      running_ = false;
+    }
+  }
 }
 
 void MessageLoopFuchsia::Terminate() {
-  loop_.Quit();
+  running_ = false;
```

Comment:
This is not sufficient. The loop may be terminated from any thread. So after setting this flag, the loop needs to woken up somehow. Otherwise, it will only terminate when the next task (unrelated and may never happen) is serviced. Let's add a unit-test that only creates a loop impl and terminates it without doing anything else. I suspect it will hang.

Reply:
`  WakeUp(fml::TimePoint::Now());` should do the trick